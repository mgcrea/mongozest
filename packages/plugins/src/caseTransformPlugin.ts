// @docs https://docs.mongodb.com/manual/reference/operator/query/type/#document-type-available-types
import '@mongozest/core';
import {isString, toString, lowerCase, upperFirst, startCase, isUndefined} from 'lodash';
import {Model, mapPathValues, DefaultSchema} from '@mongozest/core';
import {OptionalId} from 'mongodb';

type CaseTransformOperation = 'toString' | 'lowerCase' | 'upperFirst' | 'startCase';

declare module '@mongozest/core' {
  interface JsonSchemaProperty<TProp = any> {
    caseTransform?: CaseTransformOperation[];
  }
}

const transformCase = (value: unknown, caseTransformConfig: CaseTransformOperation[]): any => {
  return caseTransformConfig.reduce<string>((valueSoFar, caseTransform) => {
    switch (caseTransform) {
      case 'toString':
        return toString(valueSoFar);
      case 'lowerCase':
        return lowerCase(valueSoFar);
      case 'upperFirst':
        return upperFirst(valueSoFar);
      case 'startCase':
        return startCase(valueSoFar);
      default:
        return valueSoFar;
    }
  }, toString(value));
};

// Helper recursively parsing schema to find path where values should be casted
export const caseTransformPlugin = <TSchema extends DefaultSchema>(model: Model<TSchema>): void => {
  const caseTransformProperties = new Map<string, CaseTransformOperation[]>();
  model.post('initialize:property', (property, path) => {
    if (isString(property) || isUndefined(property.caseTransform)) {
      return;
    }
    caseTransformProperties.set(path, property.caseTransform);
  });
  model.pre('update', (_operation, _filter, update) => {
    caseTransformProperties.forEach((caseTransformConfig, path) => {
      if (update.$set) {
        mapPathValues(update.$set as OptionalId<TSchema>, path, (value) => transformCase(value, caseTransformConfig));
      }
      if (update.$push) {
        mapPathValues(update.$push as OptionalId<TSchema>, path, (value) => transformCase(value, caseTransformConfig));
      }
    });
  });
  // Handle insert
  model.pre('insert', (_operation, document) => {
    caseTransformProperties.forEach((caseTransformConfig, path) => {
      mapPathValues(document, path, (value: any) => transformCase(value, caseTransformConfig));
    });
  });
};
