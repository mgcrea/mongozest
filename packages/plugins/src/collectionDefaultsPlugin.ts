import '@mongozest/core';
import {Model, ModelConstructor, DefaultSchema} from '@mongozest/core';
import {MongoCountPreferences, OptionalId} from 'mongodb';

declare module '@mongozest/core' {
  export interface ModelConstructor<TSchema extends OptionalId<DefaultSchema> = DefaultSchema> {
    defaults?: OptionalId<TSchema>[];
  }
}

export const collectionDefaultsPlugin = <TSchema extends DefaultSchema>(
  model: Model<TSchema>,
  {maxTimeMS = 5000}: MongoCountPreferences = {}
): void => {
  // model.addStatics({defaults: []});?
  model.post('initialize', async () => {
    const {defaults} = model.constructor as ModelConstructor<TSchema>;
    // Only if the collection is empty!
    if (defaults && defaults.length) {
      const count = await model.collection.estimatedDocumentCount({}, {maxTimeMS});
      if (!count) {
        await model.insertMany(defaults as OptionalId<TSchema>[]); // @NOTE ts bug?
      }
    }
  });
};
