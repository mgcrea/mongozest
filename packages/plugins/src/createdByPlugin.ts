import {BsonType, Model, DefaultSchema} from '@mongozest/core';
import {ObjectId} from 'mongodb';

export type CreatedByPluginSchema<TProp = ObjectId> = {
  ownedBy: TProp;
  createdBy: TProp;
  updatedBy: TProp;
};

export type CreatedByPluginOptions = {
  bsonType?: BsonType;
  ref?: string;
};

export const createdByPlugin = <TSchema extends DefaultSchema & CreatedByPluginSchema>(
  model: Model<TSchema>,
  {bsonType = 'objectId', ref = 'User'}: CreatedByPluginOptions = {}
): void => {
  model.addSchemaProperties({
    ownedBy: {bsonType, ref},
    createdBy: {bsonType, ref},
    updatedBy: {bsonType, ref}
  });
};

// declare module '@mongozest/core' {
//   interface DefaultSchema {
//     ownedBy: ObjectId;
//     createdBy: ObjectId;
//     updatedBy: ObjectId;
//   }
// }
