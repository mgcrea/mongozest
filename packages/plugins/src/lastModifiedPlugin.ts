import {FilterQuery, UpdateQuery} from 'mongodb';
import {Model} from '@mongozest/core';

export interface LastModifiedPluginProps {
  updatedAt: Date;
  createdAt: Date;
}

// export interface LastModifiedPluginOptions {}

export default function lastModifiedPlugin<TSchema extends LastModifiedPluginProps>(
  model: Model<TSchema>
  // _options: LastModifiedPluginOptions
) {
  model.addSchemaProperties({
    updatedAt: {bsonType: 'date'},
    createdAt: {bsonType: 'date'}
  });
  model.pre('insert', (insert: TSchema) => {
    const currentDate = new Date();
    insert.createdAt = currentDate;
    insert.updatedAt = currentDate;
  });
  model.pre('update', (filter: FilterQuery<TSchema>, update: UpdateQuery<TSchema>) => {
    const currentDate = new Date();
    if (update.$set) {
      update.$set.updatedAt = currentDate;
    } else {
      update.$set = {updatedAt: currentDate};
    }
  });
}
