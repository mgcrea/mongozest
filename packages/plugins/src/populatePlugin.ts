// @docs https://docs.mongodb.com/manual/reference/operator/query/type/#document-type-available-types

import {get, set, isPlainObject, map, keyBy, isUndefined, isEmpty, isString} from 'lodash';
import {uniqWithObjectIds} from '@mongozest/core';
// @types
import {Model, OperationMap, mapPathValues} from '..';

import {FilterQuery} from 'mongodb';

interface FindOneOptions {
  population?: {[s: string]: number | boolean};
}

// @NOTE hybrid projection is allowed with _id field
const isExcludingProjection = (projection: {[s: string]: any}) =>
  Object.keys(projection).some(key => key !== '_id' && !projection[key]);

// Helper recursively parsing schema to find path where values should be casted
export default function autoCastingPlugin<TSchema>(model: Model<TSchema>) {
  const propsWithRefs = new Map();
  model.post('initialize:property', (prop: {[s: string]: any}, path: string) => {
    if (isString(prop) || isUndefined(prop.ref)) {
      return;
    }
    propsWithRefs.set(path, prop.ref);
  });
  // Automatically add missing keys
  model.pre('find', (filter: FilterQuery<TSchema>, options: FindOneOptions, operation: OperationMap) => {
    const {projection, population} = options;
    if (!population) {
      return;
    }
    if (!projection || isEmpty(projection) || isExcludingProjection(projection)) {
      return;
    }
    Object.keys(population).forEach(key => {
      if (get(projection, key) !== 1) {
        set(projection, key, 1);
      }
    });
  });

  model.post('findMany', async (filter: FilterQuery<TSchema>, options: FindOneOptions, operation: OperationMap) => {
    const {population} = options;
    if (!population) {
      return;
    }
    const result = operation.get('result');
    await Object.keys(population).reduce(async (soFar, key) => {
      await soFar;
      if (!propsWithRefs.has(key)) {
        return;
      }
      const ref = propsWithRefs.get(key);
      const uniqueIds = uniqWithObjectIds(map(result, key).filter(Boolean));
      const childProjectionExcludesIds = isPlainObject(population[key]) && population[key]._id === 0;
      const childProjection = isPlainObject(population[key]) ? {...population[key], _id: 1} : {};
      const resolvedChildren = await model.otherModel(ref).find({_id: {$in: uniqueIds}}, {projection: childProjection});
      const resolvedChildrenMap = keyBy(resolvedChildren, '_id');
      // Tweak references to exclude _ids
      if (childProjectionExcludesIds) {
        resolvedChildren.forEach(resolvedChild => {
          delete resolvedChild._id;
        });
      }
      // Actually populate
      result.map(doc => {
        if (doc[key]) {
          doc[key] = resolvedChildrenMap[doc[key].toString()] || null;
        }
      });
    }, Promise.resolve());
  });
  model.post('findOne', async (filter: FilterQuery<TSchema>, options: FindOneOptions, operation: OperationMap) => {
    const {population} = options;
    if (!population) {
      return;
    }
    const method = operation.get('method');
    const result = method === 'findOneAndUpdate' ? operation.get('result').value : operation.get('result');
    if (!result) {
      return;
    }
    await Object.keys(population).reduce(async (soFar, key) => {
      await soFar;
      if (!propsWithRefs.has(key)) {
        return;
      }
      const ref = propsWithRefs.get(key);
      const refValue = get(result, key);
      const isArrayValue = Array.isArray(refValue);
      const childProjectionExcludesIds = isPlainObject(population[key]) && population[key]._id === 0;
      const childProjection = isPlainObject(population[key]) ? {...population[key], _id: 1} : {};
      const resolvedChildren = await model
        .otherModel(ref)
        [isArrayValue ? 'find' : 'findOne'](
          {_id: isArrayValue ? {$in: refValue} : refValue},
          {projection: childProjection}
        );
      // Tweak references to exclude _ids
      if (childProjectionExcludesIds) {
        if (isArrayValue && Array.isArray(resolvedChildren)) {
          resolvedChildren.forEach(resolvedChild => {
            delete resolvedChild._id;
          });
        } else if (resolvedChildren) {
          delete resolvedChildren._id;
        }
      }
      // Actually populate
      set(result, key, resolvedChildren);
    }, Promise.resolve());
  });
}
