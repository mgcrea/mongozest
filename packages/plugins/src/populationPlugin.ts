import 'mongodb';
import {DefaultSchema, Model, uniqWithObjectIds} from '@mongozest/core';
import {get, isEmpty, isPlainObject, isString, isUndefined, keyBy, map, set} from 'lodash';
import {ObjectId, SchemaMember} from 'mongodb';

export type Population<T> = SchemaMember<T, number | boolean | any>;

// @NOTE hybrid projection is allowed with _id field
const isExcludingProjection = (projection: {[s: string]: any}) =>
  Object.keys(projection).some((key) => key !== '_id' && !projection[key]);

// Helper recursively parsing schema to find path where values should be casted
export const populationPlugin = <TSchema extends DefaultSchema>(model: Model<TSchema>): void => {
  const propsWithRefs = new Map();
  model.post('initialize:property', (prop: {[s: string]: any}, path: string) => {
    if (isString(prop) || isUndefined(prop.ref)) {
      return;
    }
    propsWithRefs.set(path, prop.ref);
  });
  // Automatically add missing keys
  model.pre('find', (_operation, _filter, options = {}) => {
    const {projection, population} = options;
    if (!population) {
      return;
    }
    if (!projection || isEmpty(projection) || isExcludingProjection(projection)) {
      return;
    }
    Object.keys(population).forEach((key) => {
      if (get(projection, key) !== 1) {
        set(projection, key, 1);
      }
    });
  });

  model.post('findMany', async (operation, _filter, options = {}) => {
    const {population} = options;
    if (!population) {
      return;
    }
    const result: TSchema[] = operation.get('result');
    await Object.keys(population).reduce(async (soFar, key) => {
      await soFar;
      if (!propsWithRefs.has(key)) {
        return;
      }
      const ref = propsWithRefs.get(key);
      const uniqueIds = uniqWithObjectIds(map(result, key).filter(Boolean) as ObjectId[]);
      // @ts-expect-error wtf?
      const childProjectionExcludesIds = isPlainObject(population[key]) && population[key]._id === 0;
      // @ts-expect-error wtf?
      const childProjection = isPlainObject(population[key]) ? {...population[key], _id: 1} : {};
      const resolvedChildren = await model
        .otherModel(ref)!
        .find({_id: {$in: uniqueIds}}, {projection: childProjection});
      const resolvedChildrenMap = keyBy(resolvedChildren, '_id');
      // Tweak references to exclude _ids
      if (childProjectionExcludesIds) {
        resolvedChildren.forEach((resolvedChild) => {
          delete (resolvedChild as Partial<DefaultSchema>)._id;
        });
      }
      // Actually populate
      result.map((doc) => {
        if (doc[key]) {
          Object.assign(doc, {[key]: resolvedChildrenMap[(doc[key] as ObjectId).toString()] || null});
        }
      });
    }, Promise.resolve());
  });
  model.post('findOne', async (operation, _filter, options = {}) => {
    const {population} = options;
    if (!population) {
      return;
    }
    const result = operation.get('result');
    if (!result) {
      return;
    }
    await Object.keys(population).reduce(async (soFar, key) => {
      await soFar;
      if (!propsWithRefs.has(key)) {
        return;
      }
      const ref = propsWithRefs.get(key);
      const refValue = get(result, key) as ObjectId | ObjectId[];
      const isArrayValue = Array.isArray(refValue);
      // @ts-expect-error wtf?
      const childProjectionExcludesIds = isPlainObject(population[key]) && population[key]._id === 0;
      // @ts-expect-error wtf?
      const childProjection = isPlainObject(population[key]) ? {...population[key], _id: 1} : {};

      if (isArrayValue) {
        const resolvedChildren = await model
          .otherModel(ref)!
          .find({_id: {$in: refValue as ObjectId[]}}, {projection: childProjection});
        // Tweak references to exclude _ids
        if (childProjectionExcludesIds && Array.isArray(resolvedChildren)) {
          resolvedChildren.forEach((resolvedChild) => {
            delete (resolvedChild as Partial<DefaultSchema>)._id;
          });
        }
        // Actually populate
        set(result, key, resolvedChildren);
        return;
      }

      const resolvedChild = await model
        .otherModel(ref)!
        .findOne({_id: refValue as ObjectId}, {projection: childProjection});
      if (childProjectionExcludesIds && resolvedChild) {
        delete (resolvedChild as Partial<DefaultSchema>)._id;
      }
      // Actually populate
      set(result, key, resolvedChild);
    }, Promise.resolve());
  });
};

declare module 'mongodb' {
  interface FindOneOptions<T> {
    population?: Population<T>;
  }
}
