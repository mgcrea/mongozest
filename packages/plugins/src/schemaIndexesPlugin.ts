import {find, isString, isUndefined} from 'lodash';
import {Model, log, inspect} from '@mongozest/core';

// Handle schema defaults
export default function schemaDefaultsPlugin(model: Model, {suffix = '_'} = {}) {
  const {collectionName} = model;
  const propsWithIndexes: Map<string, any> = new Map();
  model.post('initialize:property', (prop: {[s: string]: any} | string, path: string) => {
    if (isString(prop) || isUndefined(prop.index)) {
      return;
    }
    propsWithIndexes.set(path, prop.index);
  });
  // Handle document insertion
  model.post('initialize', async (doc: T) => {
    // Create indexes from static config
    const {indexes: indexesFromConfig} = model.constructor;
    if (indexesFromConfig) {
      const createdIndexesFromConfig = await indexesFromConfig.reduce(async (promiseSoFar, entry, index) => {
        const soFar = await promiseSoFar;
        const [key, options] = entry;
        const indexOptions = {name: Object.keys(key).join('_'), ...options};
        const {name} = indexOptions;
        // Do we have a named index?
        if (indexOptions.name) {
          const indexExists = await model.collection.indexExists(name);
          if (indexExists) {
            // Do we have an exact match?
            const matchingIndex = find(await model.collection.indexes(), {key, name, ...indexOptions});
            if (matchingIndex) {
              // Nothing more to do!
              return soFar;
            }
          }
        }
        log(`db.${collectionName}.createIndex(${inspect(key)}, ${inspect(indexOptions)})`);
        const indexName = await model.collection.createIndex(key, indexOptions);
        soFar.set(index, indexName);
        return soFar;
      }, Promise.resolve(new Map()));
    }

    // Create indexes from props
    // const existingIndexes = await model.collection.indexes();
    const indexesFromProps = Array.from(propsWithIndexes.entries());
    const createdIndexesFromProps = await indexesFromProps.reduce(async (promiseSoFar, entry) => {
      const soFar = await promiseSoFar;
      const [path, options] = entry;
      const indexOptions = {name: `${path}${suffix}`, ...options};
      const {name} = indexOptions;
      const key = {[path]: 1};
      // Do we have a named index?
      if (indexOptions.name) {
        const indexExists = await model.collection.indexExists(name);
        if (indexExists) {
          // Do we have an exact match?
          const matchingIndex = find(await model.collection.indexes(), {key, name, ...indexOptions});
          if (matchingIndex) {
            // Nothing more to do!
            return soFar;
          }
        }
      }
      log(`db.${collectionName}.createIndex(${inspect(key)}, ${inspect(indexOptions)})`);
      const indexName = await model.collection.createIndex(key, indexOptions);
      soFar.set(path, indexName);
      return soFar;
    }, Promise.resolve(new Map()));
  });
}
