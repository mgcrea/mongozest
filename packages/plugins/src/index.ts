import collectionDefaultsPlugin from './collectionDefaultsPlugin';
import createdByPlugin from './createdByPlugin';
import caseTransformPlugin from './caseTransformPlugin';
import populatePlugin from './populatePlugin';
import lastModifiedPlugin from './lastModifiedPlugin';
import schemaCastingPlugin from './schemaCastingPlugin';
import schemaDefaultsPlugin from './schemaDefaultsPlugin';
import schemaFakerPlugin from './schemaFakerPlugin';
import schemaIndexesPlugin from './schemaIndexesPlugin';
import schemaProjectionPlugin from './schemaProjectionPlugin';
import schemaValidationPlugin from './schemaValidationPlugin';
import trimPlugin from './trimPlugin';
import shortIdPlugin from './shortIdPlugin';

export {
  collectionDefaultsPlugin,
  caseTransformPlugin,
  createdByPlugin,
  populatePlugin,
  lastModifiedPlugin,
  schemaCastingPlugin,
  schemaDefaultsPlugin,
  schemaFakerPlugin,
  schemaIndexesPlugin,
  schemaProjectionPlugin,
  schemaValidationPlugin,
  trimPlugin,
  shortIdPlugin
};
