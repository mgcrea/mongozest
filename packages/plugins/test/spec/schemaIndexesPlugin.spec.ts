import createMongo, {JsonSchema, jsonSchemaPlugin, Model} from '@mongozest/core';
import {getDbName} from 'root/test/utils';
import {schemaIndexesPlugin} from 'src/schemaIndexesPlugin';

const DB_NAME = getDbName(__filename);

const mongo = createMongo();

type Test = {
  username?: string;
  phoneNumber?: string;
  email?: string;
};
class TestModel extends Model<Test> {
  static schema: JsonSchema<Test> = {
    username: {bsonType: 'string', index: {unique: true}},
    phoneNumber: {bsonType: 'string', index: {unique: false, name: undefined}},
    email: {bsonType: 'string', index: {unique: false, name: 'email'}}
  };
  static plugins = [jsonSchemaPlugin, schemaIndexesPlugin];
}
let testModel: TestModel;

beforeAll(async () => {
  const db = await mongo.connect(DB_NAME);
  await db.dropDatabase();
  it('should properly loadModel', async () => {
    testModel = await mongo.loadModel(TestModel);
    expect(TestModel instanceof Model).toBeTruthy();
  });
});

afterAll(async () => {
  await mongo.disconnect();
});

describe('schemaIndexesPlugin', () => {
  let testModel: Model;
  it('should properly loadModel and create indexes', async () => {
    const indexes = await testModel.collection.indexes();
    expect(indexes).toMatchSnapshot();
  });
});
