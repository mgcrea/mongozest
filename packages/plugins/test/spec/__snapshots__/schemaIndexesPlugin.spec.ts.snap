// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schemaIndexesPlugin should properly loadModel and create indexes 1`] = `
Array [
  Object {
    "key": Object {
      "_id": 1,
    },
    "name": "_id_",
    "ns": "plugins-schema-indexes-plugin-spec.test_models",
    "v": 2,
  },
  Object {
    "key": Object {
      "username": 1,
    },
    "name": "username_",
    "ns": "plugins-schema-indexes-plugin-spec.test_models",
    "unique": true,
    "v": 2,
  },
  Object {
    "key": Object {
      "phoneNumber": 1,
    },
    "name": "phoneNumber_1",
    "ns": "plugins-schema-indexes-plugin-spec.test_models",
    "v": 2,
  },
  Object {
    "key": Object {
      "email": 1,
    },
    "name": "email",
    "ns": "plugins-schema-indexes-plugin-spec.test_models",
    "v": 2,
  },
]
`;
